cli/src/cmd/deploy.rs

```diff
@@ -0,0 +1,172 @@
+    let mut data = Vec::new();
+    let mut file = fs::File::open(&contract_wasm_path)
+        .map_err(|e| format!("Failed to open {}: {}", contract_wasm_path.display(), e))?;
+    file.read_to_end(&mut data)?;
+
+    return Ok(data)
+}
+
+/// If the extrinsic was successfully included in a block, attempt to extract the code hash from the
+/// `Contracts::CodeStored` event
+fn extract_code_hash(extrinsic_result: subxt::ExtrinsicSuccess<Runtime>) -> Result<H256> {
+    match extrinsic_result.find_event::<H256>("Contracts", "CodeStored") {
+        Some(Ok(hash)) => Ok(hash),
+        Some(Err(err)) => Err(format!("Failed to decode code hash: {}", err).into()),
+        None => Err("Failed to find Contracts::CodeStored Event".into()),
+    }
+}
+
+/// Define a Runtime to provide a subset of types for interacting with the target chain via RPC.
+///
+/// # Note
+///
+/// The concrete types MUST be compatible with the target chain's respective concrete type
+/// definitions.
+#[derive(Debug, Clone, Eq, PartialEq)]
+struct Runtime;
+
+impl System for Runtime {
+    type Index = u32;
+    type BlockNumber = u32;
+    type Hash = substrate_primitives::H256;
+    type Hashing = BlakeTwo256;
+    type AccountId = <AnySignature as Verify>::Signer;
+    type Address = srml_indices::address::Address<Self::AccountId, u32>;
+    type Header = Header<Self::BlockNumber, BlakeTwo256>;
+}
+
+impl Balances for Runtime {
+    type Balance = u64;
+}
+
+impl Contracts for Runtime {}
+
+/// Put contract code to a smart contract enabled substrate chain.
+/// Returns the code hash of the deployed contract if successful.
+///
+/// Optionally supply the contract wasm path, defaults to destination contract file inferred from
+/// Cargo.toml of the current contract project.
+///
+/// Creates an extrinsic with the `Contracts::put_code` Call, submits via RPC, then waits for
+/// the `ContractsEvent::CodeStored` event.
+pub(crate) fn execute_deploy(
+    url: url::Url,
+    surl: &str,
+    password: Option<&str>,
+    gas: u64,
+    contract_wasm_path: Option<&PathBuf>,
+) -> Result<String> {
+    let signer = sr25519::Pair::from_string(surl, password)?;
+
+    let code = load_contract_code(contract_wasm_path)?;
+
+    let fut = subxt::ClientBuilder::<Runtime>::new()
+        .set_url(url)
+        .build()
+        .and_then(|cli| cli.xt(signer, None))
+        .and_then(move |xt| {
+            xt.contracts(|call| call.put_code(gas, code))
+                .submit_and_watch()
+        });
+
+    let mut rt = tokio::runtime::Runtime::new()?;
+    let extrinsic_success = rt.block_on(fut)?;
+
+    log::debug!("Deploy success: {:?}", extrinsic_success);
+
+    let code_hash = extract_code_hash(extrinsic_success)?;
+    Ok(format!("Code hash: {:?}", code_hash))
+}
+
+#[cfg(test)]
+mod tests {
+    use assert_matches::assert_matches;
+    use std::{
+        fs,
+        io::Write,
+        path,
+    };
+
+    #[test]
+    #[ignore] // depends on a local substrate node running
```

Comment:
would it make more sense to ignore this conditionally?

Reply:
You mean e.g. `#[cfg(feature="substrate-node-integration")]`?

Reply:
yeah for example if that works

Reply:
Hmm if I do this it will be run on CI with `test --all-features` (and fail)